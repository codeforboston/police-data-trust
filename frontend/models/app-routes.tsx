export enum AppRoutes {
  CONTRIBUTOR = "/contributor",
  DASHBOARD = "/search",
  FORGOT = "/forgot",
  RESET = "/reset",
  LOGIN = "/login",
  PASSPORT = "/passport",
  PROFILE = "/profile",
  REGISTER = "/register",
  OFFICER = "/officer"
}
