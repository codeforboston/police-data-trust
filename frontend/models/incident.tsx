import { Column } from "react-table"
import { GreaterThanButton } from "../shared-components/icon-buttons/icon-buttons"

export interface IncidentRecordType {
  officers: string[]
  incidentType: string
  useOfForce: string[]
  source?: string
}

export const incidentResultsColumns: Column<any>[] = [
  {
    Header: "Officers involved",
    accessor: (row: any) => (Array.isArray(row["officers"]) ? row["officers"].join(", ") : ""),
    id: "officersInvolved"
  },
  {
    Header: "Incident Type",
    accessor: (row: any) => row["incidentType"],
    id: "incidentType"
  },
  {
    Header: "Use of Force",
    accessor: (row: any) => (Array.isArray(row["useOfForce"]) ? row["useOfForce"].join(", ") : ""),
    id: "useOfForce"
  },
  {
    Header: "Source",
    accessor: "source",
    id: "source"
  },
  {
    Header: "View Record",
    Cell: () => {
      return <GreaterThanButton title={"View Record"} onclick={() => console.log("clicked")} />
    },
    id: "record"
  }
]
