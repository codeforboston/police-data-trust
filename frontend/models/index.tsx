export * from "./app-routes"
export * from "./enrollment-cta"
export * from "./info-tooltip"
export * from "./logo-sizes"
export * from "./password-aid"
export * from "./primary-input"
export * from "./secondary-input"
export * from "./response"
export * from "./results-alert"
export * from "./saved-table"
export * from "./state-select"
export * from "./toggle-options"
export * from "./visualizations"
