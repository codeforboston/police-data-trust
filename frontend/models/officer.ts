import { Incident, Perpetrator } from "../helpers/api"

export interface AgencyType {
  agencyName: string
  agencyImage: string
  agencyHqAddress: string
  websiteUrl: string
}

export const agencyColumns = [
  {
    Header: "agency",
    accessor: "agency"
  },
  {
    Header: "Address",
    accessor: "agencyHqAddress"
  },
  {
    Header: "Website",
    accessor: "websiteUrl"
  },
  {
    Header: "Agency Address",
    accessor: "agencyHqAddress"
  }
]

export interface EmploymentType {
  agency: AgencyType
  currentlyEmployed: boolean
  earliestEmployment: Date
  latestEmployment: Date
  badgeNumber?: string
}

export interface OfficerRecordType {
  recordId: number
  firstName: string
  lastName: string
  dateOfBirth?: Date
  gender?: string
  race?: string
  ethnicity?: string
  knownEmployers?: AgencyType[]
  workHistory?: EmploymentType[]
  accusations?: Perpetrator[]
  affiliations?: OfficerRecordType[]
}
