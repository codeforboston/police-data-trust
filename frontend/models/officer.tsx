import { Column } from "react-table"
import { Incident, Perpetrator } from "../helpers/api"
import { CirclePlusButton } from "../shared-components/icon-buttons/icon-buttons"
import { InfoTooltip } from "../shared-components"
import { TooltipIcons, TooltipTypes } from "./info-tooltip"

export interface AgencyType {
  agencyName: string
  agencyImage: string
  agencyHqAddress: string
  websiteUrl: string
}

export const agencyColumns = [
  {
    Header: "agency",
    accessor: "agency"
  },
  {
    Header: "Address",
    accessor: "agencyHqAddress"
  },
  {
    Header: "Website",
    accessor: "websiteUrl"
  },
  {
    Header: "Agency Address",
    accessor: "agencyHqAddress"
  }
]

export interface EmploymentType {
  agency: AgencyType
  currentlyEmployed: boolean
  earliestEmployment: Date
  latestEmployment: Date
  badgeNumber?: string
}

export interface OfficerRecordType {
  recordId: number
  firstName: string
  lastName: string
  dateOfBirth?: Date
  gender?: string
  race?: string
  ethnicity?: string
  knownEmployers?: AgencyType[]
  workHistory?: EmploymentType[]
  accusations?: Perpetrator[]
  affiliations?: OfficerRecordType[]
}

export const officerResultsColumns: Column<any>[] = [
  {
    Header: "Name",
    accessor: (row: any) => `${row["first_name"]}, ${row["last_name"].charAt(0)}`,
    id: "name"
  },
  {
    Header: () => (
      <span>
        Allegations
        <InfoTooltip type={TooltipTypes.DATETIME} icon={TooltipIcons.INFO} iconSize="xs" />
      </span>
    ),
    accessor: "allegations",
    id: "allegations"
  },
  {
    Header: "Race",
    accessor: "race",
    id: "race"
  },
  {
    Header: "Gender",
    accessor: "gender",
    id: "gender"
  },
  {
    Header: () => (
      <span>
        Rank
        <InfoTooltip type={TooltipTypes.DATETIME} icon={TooltipIcons.INFO} iconSize="xs" />
      </span>
    ),
    accessor: "rank",
    id: "rank"
  },
  {
    Header: "Employers",
    accessor: "employers",
    id: "employers"
  },
  {
    Header: "Save",
    accessor: "save",
    Cell: () => {
      return <CirclePlusButton title={"Save"} onclick={() => console.log("clicked")} />
    },
    id: "save"
  }
]
