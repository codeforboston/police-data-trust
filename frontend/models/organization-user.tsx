import { Column } from "react-table"

enum MemberRole {
  ADMIN = "Administrator",
  PUBLISHER = "Publisher",
  MEMBER = "Member",
  SUBSCRIBER = "Subscriber",
  NONE = ""
}

enum ActionOptions {
  REMOVE = "Remove",
  WITHDRAW = "Withdraw",
  INVITATION = "Invitation"
}

enum Organizations {
  ORG1 = "Organization 1",
  ORG2 = "Organization 2",
  ORG3 = "Organization 3"
}

enum Users {
  USER1 = "User 1",
  USER2 = "User 2"
}

enum Status {
  PENDING = "Pending",
  CURRENT = "Current Member"
}

export interface CurrentOrgUsers {
  select: boolean
  user: string
  role: MemberRole
  status: Status
  action: string
  id: number
}

export const orgUsers: CurrentOrgUsers[] = [
  {
    select: false,
    user: "Angel Z",
    role: MemberRole.PUBLISHER,
    status: Status.PENDING,
    action: "",
    id: 0
  },
  {
    select: false,
    user: "",
    role: MemberRole.NONE,
    status: Status.PENDING,
    action: "",
    id: 1
  },
  {
    select: false,
    user: "",
    role: MemberRole.NONE,
    status: Status.CURRENT,
    action: "",
    id: 2
  },
  {
    select: false,
    user: "",
    role: MemberRole.NONE,
    status: Status.CURRENT,
    action: "",
    id: 3
  },
  {
    select: false,
    user: "",
    role: MemberRole.NONE,
    status: Status.CURRENT,
    action: "",
    id: 4
  },
  {
    select: false,
    user: "",
    role: MemberRole.NONE,
    status: Status.CURRENT,
    action: "",
    id: 5
  },
  {
    select: false,
    user: "",
    role: MemberRole.NONE,
    status: Status.CURRENT,
    action: "",
    id: 6
  }
]

export const actionOptions = [
  {
    item: ActionOptions.REMOVE,
    text: "Remove"
  },
  {
    item: ActionOptions.WITHDRAW,
    text: "Withdraw"
  },
  {
    item: ActionOptions.INVITATION,
    text: "Invitation"
  }
]

export const changeRoleOptions = [
  {
    item: MemberRole.ADMIN,
    text: "Admin"
  },
  {
    item: MemberRole.MEMBER,
    text: "Withdraw"
  },
  {
    item: MemberRole.PUBLISHER,
    text: "Publisher"
  },
  {
    item: MemberRole.SUBSCRIBER,
    text: "Subscriber"
  },
  {
    item: MemberRole.NONE,
    text: "None"
  }
]

export const organizationOptions = [
  {
    item: Organizations.ORG1,
    text: "Organization 1"
  },
  {
    item: Organizations.ORG2,
    text: "Organization 2"
  },
  {
    item: Organizations.ORG3,
    text: "Organization 3"
  }
]

export const userOptions = [
  {
    item: Users.USER1,
    text: "User 1"
  },
  {
    item: Users.USER2,
    text: "User 2"
  }
]
