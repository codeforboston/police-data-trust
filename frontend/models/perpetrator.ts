import { Incident } from "../helpers/api"

export interface AgencyType {
  agencyName: string
  agencyImage: string
  agencyHqAddress: string
  websiteUrl: string
}

export const departmentColumns = [
  {
    Header: "Department",
    accessor: "dept"
  },
  {
    Header: "Address",
    accessor: "deptAddress"
  },
  {
    Header: "Website",
    accessor: "webAddress"
  },
  {
    Header: "Dept Address",
    accessor: "deptAddress"
  }
]

export interface PerpetratorRecordType {
  recordId: number
  firstName?: string
  lastName?: string
  badge?: string
  rank?: string
  gender?: string
  race?: string
  ethnicity?: string
  incident?: Incident
}
