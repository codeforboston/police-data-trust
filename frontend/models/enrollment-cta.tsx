import { AppRoutes } from "../models"

interface CallToActionText {
  description?: string
  linkText: string
  linkPath: AppRoutes
}

export enum CallToActionTypes {
  LOGIN = "login",
  CONTRIBUTOR = "contributor",
  REGISTER = "register",
  DASHBOARD = "dashboard",
  FORGOT = "forgot",
  REMEMBER = "remember"
}

export const enrollmentCallToActionText: { [key in CallToActionTypes]: CallToActionText } = {
  [CallToActionTypes.LOGIN]: {
    description: "Do you already have an account with us?",
    linkText: "Login instead",
    linkPath: AppRoutes.LOGIN
  },
  [CallToActionTypes.REGISTER]: {
    description: "New to the National Police Data Coalition?",
    linkText: "Register for an account",
    linkPath: AppRoutes.REGISTER
  },
  [CallToActionTypes.DASHBOARD]: {
    description: "Is the publically available data sufficient for your needs?",
    linkText: "Return to dashboard",
    linkPath: AppRoutes.DASHBOARD
  },
  [CallToActionTypes.CONTRIBUTOR]: {
    description: "Want to report an issue the police in your area?",
    linkText: "Become a Contributor",
    linkPath: AppRoutes.CONTRIBUTOR
  },
  [CallToActionTypes.FORGOT]: {
    //description: "New to the National Police Data Coalition?",
    linkText: "Forgot your password?",
    linkPath: AppRoutes.FORGOT
  },
  [CallToActionTypes.REMEMBER]: {
    description: "Do you remember you login credentials?",
    linkText: "Return to login",
    linkPath: AppRoutes.LOGIN
  }
}
