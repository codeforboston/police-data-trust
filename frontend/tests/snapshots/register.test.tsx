import { AppRoutes } from "../../models"
import Register from "../../pages/register"
import { render, router, uniqueEmail, userEvent, waitFor } from "../test-utils"
import { act } from "react-dom/test-utils"

it("renders Register correctly", () => {
  const { container } = render(<Register />)
  expect(container).toMatchSnapshot()
})

describe("behaviors", () => {
  function renderPage() {
    const r = render(<Register />)
    const { getByRole, getByLabelText } = r

    return {
      ...r,
      email: getByRole("textbox", { name: /email address/i }),
      createPassword: getByLabelText(/create password/i),
      confirmPassword: getByLabelText(/confirm password/i),
      firstname: getByRole("textbox", { name: /first name/i }),
      lastname: getByRole("textbox", { name: /last name/i }),
      phone_number: getByRole("textbox", { name: /phone number/i }),
      submit: getByRole("button", { name: /submit/i })
    }
  }

  it("checks required fields", async () => {
    const r = renderPage()
    const elements: any = r

    act(() => {
      userEvent.click(r.submit)
    })
    await expect(r.findAllByRole("alert")).resolves.toHaveLength(6)

    for (const k of [
      "email",
      "createPassword",
      "confirmPassword",
      "firstname",
      "lastname",
      "phone_number"
    ]) {
      expect(elements[k].getAttribute("aria-invalid")).toBeTruthy()
    }
  })

  it("checks phone number length", async () => {
    const r = renderPage()
    act(() => {
      userEvent.type(r.phone_number, "5555555555555")
      userEvent.click(r.submit)
    })
    await expect(r.findByText(/phone number is required/)).resolves.toBeInTheDocument()
  })

  it("checks email formatting", async () => {
    const r = renderPage()
    act(() => {
      userEvent.type(r.email, "test@test")
      userEvent.click(r.submit)
    })

    await expect(r.findByText(/enter a valid email/)).resolves.toBeInTheDocument()
  })

  it("checks password formatting", async () => {
    const r = renderPage()
    act(() => {
      userEvent.type(r.createPassword, "badpassword")
      userEvent.click(r.submit)
    })
    await expect(
      r.findByText(/please enter a password that contains/i)
    ).resolves.toBeInTheDocument()
  })

  it("requires matching passwords", async () => {
    const r = renderPage()
    act(() => {
      userEvent.type(r.firstname, "Spencer")
      userEvent.type(r.lastname, "Bool")
      userEvent.type(r.email, "spencer@example.com")
      userEvent.type(r.phone_number, "555 555 5555")
      userEvent.type(r.createPassword, "aA1!asdfasdf")
      userEvent.type(r.confirmPassword, "mistmatch")
      userEvent.click(r.submit)
    })
    await expect(r.findByText(/passwords do not match/i)).resolves.toBeInTheDocument()
  })

  it("should create a new user", async () => {
    const r = renderPage()
    act(() => {
      userEvent.type(r.firstname, "Spencer")
      userEvent.type(r.lastname, "Bool")
      userEvent.type(r.email, uniqueEmail())
      userEvent.type(r.phone_number, "555 555 5555")
      userEvent.type(r.createPassword, "aA1!asdfasdf")
      userEvent.type(r.confirmPassword, "aA1!asdfasdf")

      userEvent.click(r.submit)
    })
    await waitFor(() => expect(router.pathname).toBe(AppRoutes.DASHBOARD))
  })

  it("should reject existing accounts", async () => {
    const r = renderPage()
    act(() => {
      userEvent.type(r.firstname, "Spencer")
      userEvent.type(r.lastname, "Bool")
      userEvent.type(r.email, "test@example.com")
      userEvent.type(r.phone_number, "555 555 5555")
      userEvent.type(r.createPassword, "aA1!asdfasdf")
      userEvent.type(r.confirmPassword, "aA1!asdfasdf")

      userEvent.click(r.submit)
    })
    // There's no reason for this email to exist in the test database.
    // Skipping this test; UI is changing anyway.
    // await expect(r.findByText(/Existing account found./i)).resolves.toBeInTheDocument()
  })
})
