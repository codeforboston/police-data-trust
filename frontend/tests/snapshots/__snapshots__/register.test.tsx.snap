// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Register correctly 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
  <div>
    <section
      class="enrollmentSection"
    >
      <div
        class="header"
      >
        <img
          alt="A blue circular logo design reads 'National Police Data Coalition'"
          class="logo medium"
          src="/NPDCLogo.svg"
        />
        <h1>
          Register: Viewer Account
          <div
            aria-describedby="viewerAccountsTooltip"
            aria-label="Learn more about Viewer Accounts"
            class="tooltipContainer"
            role="tooltip"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-question-circle fa-w-16 fa-sm tooltipIcon"
              data-icon="question-circle"
              data-prefix="far"
              data-tooltip-content="A Viewer account allows you to see data that is publically available.:::::Once registered, you may optionally apply to see legally protected data through a Passport account"
              data-tooltip-id="table-column-tooltip"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                fill="currentColor"
              />
            </svg>
          </div>
        </h1>
      </div>
      <form>
        <fieldset
          class="inputLine"
        >
          <div
            class="defaultInputContainer inputContainer undefined"
          >
            <label
              for="firstNameInput"
            >
              First Name:
            </label>
            <input
              aria-invalid="false"
              aria-required="true"
              class="inputField"
              id="firstNameInput"
              name="firstName"
              type="text"
              value=""
            />
          </div>
          <div
            class="defaultInputContainer inputContainer undefined"
          >
            <label
              for="lastNameInput"
            >
              Last Name:
            </label>
            <input
              aria-invalid="false"
              aria-required="true"
              class="inputField"
              id="lastNameInput"
              name="lastName"
              type="text"
              value=""
            />
          </div>
        </fieldset>
        <fieldset
          class="inputLine"
        >
          <div
            class="defaultInputContainer inputContainer undefined"
          >
            <label
              for="emailAddressInput"
            >
              Email Address:
            </label>
            <input
              aria-invalid="false"
              aria-required="true"
              class="inputField"
              id="emailAddressInput"
              name="emailAddress"
              type="email"
              value=""
            />
          </div>
          <div
            class="defaultInputContainer inputContainer undefined"
          >
            <label
              for="phoneNumberInput"
            >
              Phone Number:
            </label>
            <input
              aria-invalid="false"
              aria-required="true"
              class="inputField"
              id="phoneNumberInput"
              name="phoneNumber"
              type="tel"
              value=""
            />
          </div>
        </fieldset>
        <fieldset
          aria-describedby="passwordAid"
          class="inputLine"
        >
          <div
            class="defaultInputContainer inputContainer undefined"
          >
            <label
              for="createPasswordInput"
            >
              Create Password:
            </label>
            <input
              aria-invalid="false"
              aria-required="true"
              class="inputField"
              id="createPasswordInput"
              name="createPassword"
              type="password"
              value=""
            />
          </div>
          <div
            class="defaultInputContainer inputContainer undefined"
          >
            <label
              for="confirmPasswordInput"
            >
              Confirm Password:
            </label>
            <input
              aria-invalid="false"
              aria-required="true"
              class="inputField"
              id="confirmPasswordInput"
              name="confirmPassword"
              type="password"
              value=""
            />
          </div>
        </fieldset>
        <div
          class="passwordAid"
        >
          <p
            id="passwordAid"
          >
            Use eight or more characters with a mix of letters, numbers, and symbols
          </p>
          <button
            aria-checked="true"
            class="passwordToggle"
            role="switch"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-eye fa-w-18 "
              data-icon="eye"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
                fill="currentColor"
              />
            </svg>
             
            Show password
          </button>
        </div>
        <button
          class="primaryButton"
          type="submit"
        >
          Submit
        </button>
      </form>
      <div
        class="callToActionContainer"
      >
        <p>
          Do you already have an account with us?
        </p>
        <a
          class="callToActionLink"
          href="/login"
        >
          Login instead
        </a>
      </div>
    </section>
  </div>
</div>
`;
